Make configuration values mutable during runtime, Part 1

klaren opened the pull request:
This is a preparatory PR for making the settings mutable during runtime. 

- Settings and configuration API have been reduced to a smaller surface to simplify thread-safety.
- `Config` is now augmented instead of copied everywhere, thus everyone refers to the same object, to allow a single source of configuration values that can be polled or subscribed to at a later stage.
- `Setting` has a new property, `dynamic`, that indicates whether it can be changed at runtime or not. 

Three settings have been changed to use polling, namely:
- dbms.transaction.timeout
- dbms.logs.query.threshold
- dbms.logs.query.enabled

When changed it will only affect newly created transaction and queries. Even though this is the case, there is not yet any way of changing the actual values from the client side and this will come in part 2 of this feature. This PR is proposed to be merged now due to the massive changeset.

This PR breaks some tools that have to be fixed after this is merged, separate PRs for affected projects is already in place.

klaren committed:
Introduced isReloadable to settings api to reflect the fact that it can be reloaded during runtime.
Refactor Settings to use a builder instead to cope with all method overloads.

klaren committed:
Removed Config.empty() since it's exactly the same as Config.defaults(),
to avoid confusions as to why the configuration is not empty.

klaren committed:
Refactor Config to use a builder instead.

klaren committed:
Replace Config.with() with Config.augment()

klaren committed:
Fixed order of overridden defaults

klaren committed:
Fixed failing test due to bolt advertised port not resolving correctly

klaren committed:
Added missing annotations to CopyOnWriteHashMap

klaren committed:
Deprecated Setting.from(Configuration) method in favor or Configuration.get(Setting).
Added convenient methods for Config.augment() and Config.defaults() and refactor code to use those.
Refactored ConfigurationValidator to actually use the Config instead of a Map<String,String>.

klaren committed:
Removed unnecessary stringMap boxing

klaren committed:
Replaced hard coded filename with Config.DEFAULT_CONFIG_FILE_NAME

klaren committed:
Deprecated settings API for complete overhaul in 4.0
Renamed isReloadable to Dynamic and maid it an annotation instead

tinwelint reviewed (approved):
There's a lot of changes and whereas it feels hard to fully get into every little detail it looks proper and in the right direction

tinwelint commented:
@klaren a rebase would be in order too

MishaDemianenko commented on community/neo4j/pom.xml:
```diff
@@ -134,6 +134,11 @@
+    <dependency>
+      <groupId>com.google.code.findbugs</groupId>
+      <artifactId>annotations</artifactId>
+      <scope>test</scope>
```
not sure why scope of this dependency is different compare with other findbugs annotations?

klaren replied:
Intellij automatically added that, removed!

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/OnlineBackupCommand.java:
```diff
@@ -296,19 +293,11 @@ private String renameExistingBackup( final Path folder, final String oldName ) t
-                        .loadConfigWithConnectorsDisabled(
-                                Optional.of( homeDir.toFile() ),
-                                Optional.of( configDir.resolve( "neo4j.conf" ).toFile() ) ) );
+                Config.fromFile( configDir.resolve( "neo4j.conf" ) ).withHome( homeDir )
```
Config class have constant defined for default config file name

klaren replied:
All hard coded strings have been replaced now.

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/backup/OnlineBackupSettings.java:
```diff
@@ -36,9 +36,9 @@
 {
     @Description( "Enable support for running online backups" )
-    public static final Setting<Boolean> online_backup_enabled = setting( "dbms.backup.enabled", BOOLEAN, TRUE );
+    public static final Setting<Boolean> online_backup_enabled = setting( "dbms.backup.enabled", BOOLEAN, TRUE ).build();
```
I do not think we need to add those 'build()' part everywhere settings are defined right?

MishaDemianenko commented on enterprise/backup/src/main/java/org/neo4j/restore/RestoreDatabaseCli.java:
```diff
@@ -55,10 +51,10 @@ public RestoreDatabaseCli( Path homeDir, Path configDir )
-                Optional.of( configDir.resolve( "neo4j.conf" ).toFile() ) );
-
-        return config.with( stringMap( DatabaseManagementSystemSettings.active_database.name(), databaseName ) );
+        return Config.fromFile( configDir.resolve( "neo4j.conf" ) )
```
default config file name is defined in Config class already as a constant

MishaDemianenko commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/CausalClusterConfigurationValidator.java:
```diff
@@ -38,41 +36,36 @@
-        validateLoadBalancing( rawConfig, log );
-
-        return rawConfig;
+        return Collections.emptyMap();
```
not sure why we need to return empty map now at all

klaren replied:
Config validators are allowed to produce new values for the config. This behavior is questionable and should probably be removed in the future. We'll leave it for now since public API restricts refactoring.

MishaDemianenko commented on enterprise/causal-clustering/src/main/java/org/neo4j/commandline/dbms/UnbindFromClusterCommand.java:
```diff
@@ -65,16 +60,9 @@ static Arguments arguments()
-    private static List<Class<?>> settings()
-    {
-        return Arrays.asList( GraphDatabaseSettings.class, DatabaseManagementSystemSettings.class );
+        return fromFile( configDir.resolve( "neo4j.conf" ) )
```
Constant for default config file name already defined in Config class

MishaDemianenko commented on enterprise/cluster/src/main/java/org/neo4j/configuration/HaConfigurationValidator.java:
```diff
@@ -39,36 +38,30 @@
-        validateInitialHosts( rawConfig::get );
-
-        return rawConfig;
+        return Collections.emptyMap();
```
here empty map as result always as well

MishaDemianenko commented on community/dbms/src/main/java/org/neo4j/commandline/dbms/DumpCommand.java:
```diff
@@ -101,12 +99,11 @@ public void execute( String[] args ) throws IncorrectUsage, CommandFailed
-                        Optional.of( configDir.resolve( "neo4j.conf" ).toFile() ) )
-                .with( stringMap( DatabaseManagementSystemSettings.active_database.name(), databaseName ) )
-                .get( database_path ).toPath();
+        return Config.fromFile( configDir.resolve( "neo4j.conf" ) )
```
please use already defined constant for default config file name

MishaDemianenko commented on community/dbms/src/main/java/org/neo4j/commandline/dbms/ImportCommand.java:
```diff
@@ -287,9 +286,10 @@ public void execute( String[] args ) throws IncorrectUsage, CommandFailed
-                Optional.of( configDir.resolve( "neo4j.conf" ).toFile() ) );
-        additionalConfig.put( DatabaseManagementSystemSettings.active_database.name(), databaseName );
-        return config.with( additionalConfig );
+        return Config.fromFile( configDir.resolve( "neo4j.conf" ) )
```
config name here as well

MishaDemianenko commented on enterprise/kernel/src/main/java/org/neo4j/graphdb/factory/EnterpriseGraphDatabaseFactory.java:
```diff
@@ -45,14 +45,15 @@
             @Override
             public GraphDatabaseService newDatabase( Config config )
             {
+                config.augment( "unsupported.dbms.ephemeral", FALSE );
```
we can also update code to reference setting instead of string here?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/ssl/SslPolicyLoader.java:
```diff
@@ -165,8 +161,7 @@ private SslPolicy loadOrCreateLegacyPolicy()
     {
-        Function<ConfigValues,Stream<String>> enumeration = GroupSettingSupport.enumerate( SslPolicyConfig.class );
-        Set<String> policyNames = config.view( enumeration ).collect( toSet() );
+        Set<String> policyNames = config.identifiersFromPrefix( SslPolicyConfig.class.getAnnotation( Group.class ).value() );
```
it looks like it would be more clear if we will pass only class that has Group annotation and will do the extraction of annotation value internally and not sure 'identifiersFromPrefix' is really describing what we doing here with group of settings

klaren replied:
Agree, updated

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -604,10 +793,25 @@ private void warnAboutDeprecations( Map<String,String> userSettings )
+    @Override
+    public String toString()
+    {
+        List<String> keys = new ArrayList<>( params.keySet() );
```
this can be done simpler using a sorted stream of param entries to produce output string

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -74,237 +73,467 @@
     private final ConfigurationMigrator migrator;
-    private final Optional<File> configFile;
     private final List<ConfigurationValidator> validators = new ArrayList<>();
+    private final Map<String,String> overriddenDefaults = new CopyOnWriteHashMap<>(); // TODO : has bo be reapplied every update
```
todo

klaren replied:
Todo for part 2 of this PR, already gone on that branch

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -74,237 +73,467 @@
     private final List<ConfigOptions> configOptions;
 
-    private final Map<String,String> params = new ConcurrentHashMap<>();
+    private final Map<String,String> params = new CopyOnWriteHashMap<>();
```
why do we use COW map instead of for example CHM?

klaren replied:
To avoid locks for read heavy workloads

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java:
```diff
@@ -74,237 +73,467 @@
+     * @return A set of uniquely and sorted group keys.
      */
-    public Config withDefaults( Map<String,String> additionalDefaults )
+    public Set<String> identifiersFromPrefix( String prefix )
```
if this is prefix base search why we can't simply use start with?

klaren replied:
We use the match to extract the identifier. Javadoc is updated to make this more clear.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/configuration/IndividualSettingsValidator.java:
```diff
@@ -102,6 +94,6 @@ public IndividualSettingsValidator( boolean warnOnUnknownSettings )
         } );
 
-        return validConfig;
+        return Collections.emptyMap();
```
empty map as result again

MishaDemianenko reviewed (changes requested):
Couple of small comments and questions otherwise looks good!

klaren committed:
Removed initllij auto-added scope

klaren commented:
@MishaDemianenko @tinwelint thanks for comments! New commits available addressing them.

klaren committed:
Removed hard coded config filenames in favor of constant in Config.
Renamed setting to buildSetting when additional parameters are needed to make it clear that a builder is returned.
Refactored identifiersFromPrefix to to use the original annotation Group, that got lost in previous refactoring.
